### PR TITLE
Razkalla/HW02

### PR DESCRIPTION
**Pull request by AbanoubSeddik**

### CODE REVIEW

File: Main.hs [hunk 1]
```diff
@@ -9,30 +9,39 @@ module Main where
...
-productOfDifference = undefined
+productOfDifference xs@(_:_:_)= product (zipWith (-) xs (tail xs))
+productOfDifference _ = error "List too short"
+
 
 isSorted :: [Int] -> Bool 
-isSorted xs = undefined
+isSorted [] = True 
+isSorted [_] = True
+isSorted (x:y:xs) = x <= y && isSorted (y:xs)
```
Comment (alex28sh) [inline Main.hs:28]: What about using higher-order functions?

File: Main.hs [hunk 2]
```diff
@@ -9,30 +9,39 @@ module Main where
...
 
 countElement :: Int -> [Int] -> Int 
-countElement = undefined
+countElement x [] = 0
+countElement z (x:xs) = 
+  (if z == x then 1 else 0) + countElement z xs
```
Comment (alex28sh) [inline Main.hs:33]: again

File: Main.hs [hunk 3]
```diff
@@ -9,30 +9,39 @@ module Main where
...
 
 dotProduct :: [Int] -> [Int] -> Int 
-dotProduct = undefined
+dotProduct xs ys = sum [x * y | (x, y) <- zip xs ys]
```
Comment (alex28sh) [inline Main.hs:36]: zipWith?

File: Main.hs [hunk 4]
```diff
@@ -9,30 +9,39 @@ module Main where
...
 
 applyAll :: [a -> b] -> a -> [b]
-applyAll = undefined
+applyAll funs x = map ($ x) funs
 
 interleave :: [a] -> [a] -> [a] 
-interleave = undefined
+interleave [] _ = []
+interleave _ [] = []
+interleave (x:xs) (y:ys) = x:y: interleave xs ys
```
Comment (alex28sh) [inline Main.hs:44]: Too much code, use higher-order functions on lists

**Review by alex28sh — CHANGES_REQUESTED** [review]
4/8

**Comment by kajigor** [comment]
Good.

**Pull request closed without merging**